File: tools/benchmark_zstd.py
#!/usr/bin/env python3
"""Benchmark zstd encoders and verify decoded bytes against each fixture."""

import argparse
import csv
import statistics
import subprocess
import sys
from pathlib import Path


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--fixtures",
        type=Path,
        default=Path("/tmp/zstd-bench/fixtures"),
        help="Directory containing input fixture files.",
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path("/tmp/zstd-current-branch-bench"),
        help="Temporary directory for compressed outputs.",
    )
    parser.add_argument(
        "--current-bin",
        default="/tmp/ruzstd-cli-huffman-maxheight",
        help="Path to the current ruzstd-cli binary.",
    )
    parser.add_argument(
        "--upstream-bin",
        default="/tmp/ruzstd-cli-baseline",
        help="Path to the upstream/baseline ruzstd-cli binary.",
    )
    parser.add_argument(
        "--c-zstd-bin",
        default="/usr/bin/zstd",
        help="Path to the C zstd binary.",
    )
    parser.add_argument("-l", "--level", type=int, default=1)
    parser.add_argument("--runs", type=int, default=3)
    parser.add_argument(
        "--csv-output",
        type=Path,
        default=Path("/tmp/zstd-rs-benchmark.csv"),
    )
    parser.add_argument(
        "--md-output",
        type=Path,
        default=Path("/tmp/zstd-rs-benchmark.md"),
    )
    parser.add_argument(
        "--no-sync",
        action="store_true",
        help="Skip sync before timed runs.",
    )
    return parser.parse_args()


def encoder_commands(args):
    return [
        (
            "upstream",
            [
                args.upstream_bin,
                "compress",
                "{input}",
                "{output}",
                "-l",
                str(args.level),
            ],
        ),
        (
            "current",
            [
                args.current_bin,
                "compress",
                "{input}",
                "{output}",
                "-l",
                str(args.level),
            ],
        ),
        (
            "c_zstd",
            [
                args.c_zstd_bin,
                "-q",
                "-f",
                f"-{args.level}",
                "{input}",
                "-o",
                "{output}",
            ],
        ),
    ]


def render(command, fixture, output):
    return [part.format(input=str(fixture), output=str(output)) for part in command]


def run_timed(command, output):
    time_file = output.with_suffix(output.suffix + ".time")
    subprocess.run(
        ["/usr/bin/time", "-f", "%e\t%U\t%S\t%M", "-o", str(time_file), *command],
        check=True,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    elapsed, user, system, rss = time_file.read_text().strip().split("\t")
    time_file.unlink()
    return float(elapsed), float(user) + float(system), int(rss)


def verify_decoded_matches(zstd_bin, compressed, original):
    with original.open("rb") as original_file:
        process = subprocess.Popen(
            [zstd_bin, "-q", "-d", "-c", str(compressed)],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
        )
        assert process.stdout is not None
        while True:
            decoded = process.stdout.read(1024 * 1024)
            expected = original_file.read(len(decoded) if decoded else 1024 * 1024)
            if decoded != expected:
                process.kill()
                process.wait()
                raise RuntimeError(
                    f"decoded output did not match original: {compressed}"
                )
            if not decoded:
                break
        extra = original_file.read(1)
        returncode = process.wait()
        if returncode != 0:
            raise subprocess.CalledProcessError(returncode, process.args)
        if extra:
            raise RuntimeError(f"decoded output ended early: {compressed}")


def run_benchmarks(args):
    args.output_dir.mkdir(exist_ok=True)
    rows = []
    for fixture in sorted(args.fixtures.iterdir()):
        if not fixture.is_file():
            continue
        for encoder, command_template in encoder_commands(args):
            output = args.output_dir / f"{fixture.name}.{encoder}.zst"
            command = render(command_template, fixture, output)
            subprocess.run(
                command,
                check=True,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
            verify_decoded_matches(args.c_zstd_bin, output, fixture)
            output.unlink()

            elapsed_runs = []
            cpu_runs = []
            rss_runs = []
            size = 0
            for _ in range(args.runs):
                if not args.no_sync:
                    subprocess.run(["sync"], check=True, stdout=subprocess.DEVNULL)
                elapsed, cpu, rss = run_timed(command, output)
                verify_decoded_matches(args.c_zstd_bin, output, fixture)
                size = output.stat().st_size
                output.unlink()
                elapsed_runs.append(elapsed)
                cpu_runs.append(cpu)
                rss_runs.append(rss)

            rows.append(
                {
                    "fixture": fixture.name,
                    "encoder": encoder,
                    "bytes": str(size),
                    "elapsed": f"{statistics.median(elapsed_runs):.2f}",
                    "cpu": f"{statistics.median(cpu_runs):.2f}",
                    "rss": str(max(rss_runs)),
                }
            )
    return rows


def write_csv(path, rows):
    with path.open("w", newline="") as csv_file:
        writer = csv.DictWriter(
            csv_file,
            fieldnames=["fixture", "encoder", "bytes", "elapsed", "cpu", "rss"],
        )
        writer.writeheader()
        writer.writerows(rows)


def pct_improvement(before, after):
    if before == 0:
        return 0.0
    return (before - after) * 100.0 / before


def write_markdown(path, rows, csv_path):
    by_fixture = {}
    for row in rows:
        by_fixture.setdefault(row["fixture"], {})[row["encoder"]] = row

    headers = [
        "Fixture",
        "Upstream bytes",
        "C bytes",
        "New bytes",
        "% Improvement",
        "Upstream CPU",
        "C CPU",
        "New CPU",
        "% Improvement",
    ]
    table_rows = []
    for fixture in sorted(by_fixture):
        encoders = by_fixture[fixture]
        upstream = encoders["upstream"]
        current = encoders["current"]
        c_zstd = encoders["c_zstd"]
        upstream_bytes = int(upstream["bytes"])
        current_bytes = int(current["bytes"])
        c_bytes = int(c_zstd["bytes"])
        upstream_cpu = float(upstream["cpu"])
        current_cpu = float(current["cpu"])
        c_cpu = float(c_zstd["cpu"])
        table_rows.append(
            [
                fixture,
                f"{upstream_bytes:,}",
                f"{c_bytes:,}",
                f"{current_bytes:,}",
                f"{pct_improvement(upstream_bytes, current_bytes):+.1f}%",
                f"{upstream_cpu:.2f}s",
                f"{c_cpu:.2f}s",
                f"{current_cpu:.2f}s",
                f"{pct_improvement(upstream_cpu, current_cpu):+.1f}%",
            ]
        )

    widths = [max(len(str(item)) for item in column) for column in zip(headers, *table_rows)]

    def fmt(row):
        return "  ".join(str(value).ljust(widths[idx]) for idx, value in enumerate(row))

    lines = [
        "# zstd-rs Benchmark",
        "",
        f"Source CSV: `{csv_path}`",
        "",
        "Percent improvements compare new/current against upstream. Each compressed output is decoded with C zstd and byte-compared against the original fixture.",
        "",
        "```text",
        fmt(headers),
        fmt(["-" * width for width in widths]),
    ]
    lines.extend(fmt(row) for row in table_rows)
    lines.extend(["```", ""])
    path.write_text("\n".join(lines))


def main():
    args = parse_args()
    rows = run_benchmarks(args)
    write_csv(args.csv_output, rows)
    write_markdown(args.md_output, rows, args.csv_output)
    print(args.csv_output)
    print(args.md_output)


if __name__ == "__main__":
    try:
        main()
    except BrokenPipeError:
        sys.exit(1)

File: tools/src/bin/benchmark_zstd.rs
use std::{
    cmp::Ordering,
    env, fs, io,
    path::{Path, PathBuf},
    process::Command,
};

use zstd_rs_tools::{
    benchmark_tmp, csv_escape, has_flag, parse_value, repo_root, run_command_silent,
    verify_decoded_matches, write_all,
};

#[derive(Clone)]
struct Args {
    fixtures: PathBuf,
    output_dir: PathBuf,
    current_bin: PathBuf,
    upstream_bin: PathBuf,
    c_zstd_bin: PathBuf,
    level: u8,
    runs: usize,
    csv_output: PathBuf,
    md_output: PathBuf,
    commentary: String,
    no_sync: bool,
}

#[derive(Clone)]
struct Row {
    fixture: String,
    encoder: &'static str,
    bytes: u64,
    elapsed: f64,
    cpu: f64,
    rss: u64,
}

fn main() -> io::Result<()> {
    let args = parse_args()?;
    let rows = run_benchmarks(&args)?;
    write_csv(&args.csv_output, &rows)?;
    write_markdown(&args.md_output, &rows, &args.csv_output, &args.commentary)?;
    println!("{}", args.csv_output.display());
    println!("{}", args.md_output.display());
    Ok(())
}

fn parse_args() -> io::Result<Args> {
    let raw = env::args().skip(1).collect::<Vec<_>>();
    if has_flag(&raw, "--help") || has_flag(&raw, "-h") {
        print_help();
        std::process::exit(0);
    }
    let repo = repo_root();
    let tmp = benchmark_tmp();
    Ok(Args {
        fixtures: PathBuf::from(parse_value(
            &raw,
            "--fixtures",
            repo.join("benchmarks")
                .join("fixtures")
                .join("broad-local")
                .display()
                .to_string(),
        )),
        output_dir: PathBuf::from(parse_value(
            &raw,
            "--output-dir",
            tmp.join("benchmark-output").display().to_string(),
        )),
        current_bin: PathBuf::from(parse_value(
            &raw,
            "--current-bin",
            repo.join("target")
                .join("release")
                .join("ruzstd-cli")
                .display()
                .to_string(),
        )),
        upstream_bin: PathBuf::from(parse_value(
            &raw,
            "--upstream-bin",
            tmp.join("ruzstd-cli-baseline").display().to_string(),
        )),
        c_zstd_bin: PathBuf::from(parse_value(&raw, "--c-zstd-bin", "/usr/bin/zstd")),
        level: parse_value(&raw, "--level", parse_value(&raw, "-l", "1"))
            .parse()
            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?,
        runs: parse_value(&raw, "--runs", "3")
            .parse()
            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?,
        csv_output: PathBuf::from(parse_value(
            &raw,
            "--csv-output",
            tmp.join("zstd-rs-benchmark.csv").display().to_string(),
        )),
        md_output: PathBuf::from(parse_value(
            &raw,
            "--md-output",
            tmp.join("zstd-rs-benchmark.md").display().to_string(),
        )),
        commentary: parse_value(&raw, "--commentary", ""),
        no_sync: has_flag(&raw, "--no-sync"),
    })
}

fn print_help() {
    println!(
        "Usage: benchmark_zstd [--fixtures DIR] [--output-dir DIR] [--current-bin PATH] \\\n    [--upstream-bin PATH] [--c-zstd-bin PATH] [-l LEVEL] [--runs N] \\\n    [--csv-output PATH] [--md-output PATH] [--commentary TEXT] [--no-sync]\n\nOptions:\n  --fixtures DIR       Directory containing input fixture files.\n  --output-dir DIR     Temporary directory for compressed outputs.\n  --current-bin PATH   Path to the current ruzstd-cli binary.\n  --upstream-bin PATH  Path to the upstream/baseline ruzstd-cli binary.\n  --c-zstd-bin PATH    Path to the C zstd binary.\n  -l, --level LEVEL    Compression level.\n  --runs N             Timed runs per fixture.\n  --csv-output PATH    CSV output path.\n  --md-output PATH     Markdown output path.\n  --commentary TEXT    Short note for the Markdown output.\n  --no-sync            Skip sync before timed runs.\n  -h, --help           Show this help message."
    );
}

fn encoder_commands(args: &Args, input: &Path, output: &Path) -> Vec<(&'static str, Command)> {
    let mut upstream = Command::new(&args.upstream_bin);
    upstream
        .arg("compress")
        .arg(input)
        .arg(output)
        .arg("-l")
        .arg(args.level.to_string());

    let mut current = Command::new(&args.current_bin);
    current
        .arg("compress")
        .arg(input)
        .arg(output)
        .arg("-l")
        .arg(args.level.to_string());

    let mut c_zstd = Command::new(&args.c_zstd_bin);
    c_zstd
        .args(["-q", "-f"])
        .arg(format!("-{}", args.level))
        .arg(input)
        .arg("-o")
        .arg(output);

    vec![
        ("upstream", upstream),
        ("current", current),
        ("c_zstd", c_zstd),
    ]
}

fn run_benchmarks(args: &Args) -> io::Result<Vec<Row>> {
    fs::create_dir_all(&args.output_dir)?;
    let mut fixtures = fs::read_dir(&args.fixtures)?
        .filter_map(Result::ok)
        .map(|entry| entry.path())
        .filter(|path| path.is_file())
        .collect::<Vec<_>>();
    fixtures.sort();

    let mut rows = Vec::new();
    for fixture in fixtures {
        let fixture_name = fixture
            .file_name()
            .and_then(|name| name.to_str())
            .ok_or_else(|| io::Error::other("fixture name is not UTF-8"))?
            .to_string();
        for (encoder, mut command) in encoder_commands(args, &fixture, Path::new("{output}")) {
            let output = args
                .output_dir
                .join(format!("{fixture_name}.{encoder}.zst"));
            replace_output_arg(&mut command, &output);
            run_command_silent(&mut command)?;
            verify_decoded_matches(&args.c_zstd_bin, &output, &fixture)?;
            fs::remove_file(&output)?;

            let mut elapsed_runs = Vec::with_capacity(args.runs);
            let mut cpu_runs = Vec::with_capacity(args.runs);
            let mut rss_runs = Vec::with_capacity(args.runs);
            let mut size = 0;
            for _ in 0..args.runs {
                if !args.no_sync {
                    let mut sync = Command::new("sync");
                    run_command_silent(&mut sync)?;
                }
                let (elapsed, cpu, rss) = run_timed(&command, &output)?;
                verify_decoded_matches(&args.c_zstd_bin, &output, &fixture)?;
                size = fs::metadata(&output)?.len();
                fs::remove_file(&output)?;
                elapsed_runs.push(elapsed);
                cpu_runs.push(cpu);
                rss_runs.push(rss);
            }

            rows.push(Row {
                fixture: fixture_name.clone(),
                encoder,
                bytes: size,
                elapsed: median(&mut elapsed_runs),
                cpu: median(&mut cpu_runs),
                rss: rss_runs.into_iter().max().unwrap_or(0),
            });
        }
    }
    Ok(rows)
}

fn replace_output_arg(command: &mut Command, output: &Path) {
    let args = command
        .get_args()
        .map(|arg| {
            if arg == "{output}" {
                output.as_os_str().to_os_string()
            } else {
                arg.to_os_string()
            }
        })
        .collect::<Vec<_>>();
    let program = command.get_program().to_os_string();
    *command = Command::new(program);
    command.args(args);
}

fn run_timed(command: &Command, output: &Path) -> io::Result<(f64, f64, u64)> {
    let time_file = output.with_extension("zst.time");
    let mut timed = Command::new("/usr/bin/time");
    timed
        .args(["-f", "%e\t%U\t%S\t%M", "-o"])
        .arg(&time_file)
        .arg(command.get_program())
        .args(command.get_args());
    run_command_silent(&mut timed)?;
    let text = fs::read_to_string(&time_file)?;
    fs::remove_file(&time_file)?;
    let fields = text.trim().split('\t').collect::<Vec<_>>();
    if fields.len() != 4 {
        return Err(io::Error::other(format!("unexpected time output: {text}")));
    }
    let elapsed = fields[0].parse::<f64>().unwrap_or(0.0);
    let user = fields[1].parse::<f64>().unwrap_or(0.0);
    let system = fields[2].parse::<f64>().unwrap_or(0.0);
    let rss = fields[3].parse::<u64>().unwrap_or(0);
    Ok((elapsed, user + system, rss))
}

fn median(values: &mut [f64]) -> f64 {
    values.sort_by(|left, right| left.partial_cmp(right).unwrap_or(Ordering::Equal));
    values[values.len() / 2]
}

fn write_csv(path: &Path, rows: &[Row]) -> io::Result<()> {
    let mut csv = String::from("fixture,encoder,bytes,elapsed,cpu,rss\n");
    for row in rows {
        csv.push_str(&format!(
            "{},{},{},{:.2},{:.2},{}\n",
            csv_escape(&row.fixture),
            row.encoder,
            row.bytes,
            row.elapsed,
            row.cpu,
            row.rss
        ));
    }
    write_all(path, &csv)
}

fn pct_improvement(before: f64, after: f64) -> f64 {
    if before == 0.0 {
        0.0
    } else {
        (before - after) * 100.0 / before
    }
}

fn write_markdown(path: &Path, rows: &[Row], csv_path: &Path, commentary: &str) -> io::Result<()> {
    let mut fixtures = rows
        .iter()
        .map(|row| row.fixture.as_str())
        .collect::<Vec<_>>();
    fixtures.sort_unstable();
    fixtures.dedup();

    let headers = [
        "Fixture",
        "Upstream bytes",
        "C bytes",
        "New bytes",
        "% Improvement",
        "Upstream CPU",
        "C CPU",
        "New CPU",
        "% Improvement",
    ];
    let mut table_rows = Vec::new();
    for fixture in fixtures {
        let upstream = find_row(rows, fixture, "upstream")?;
        let current = find_row(rows, fixture, "current")?;
        let c_zstd = find_row(rows, fixture, "c_zstd")?;
        table_rows.push(vec![
            fixture.to_string(),
            format_number(upstream.bytes),
            format_number(c_zstd.bytes),
            format_number(current.bytes),
            format!(
                "{:+.1}%",
                pct_improvement(upstream.bytes as f64, current.bytes as f64)
            ),
            format!("{:.2}s", upstream.cpu),
            format!("{:.2}s", c_zstd.cpu),
            format!("{:.2}s", current.cpu),
            format!("{:+.1}%", pct_improvement(upstream.cpu, current.cpu)),
        ]);
    }

    let widths = (0..headers.len())
        .map(|column| {
            table_rows
                .iter()
                .map(|row| row[column].len())
                .chain([headers[column].len()])
                .max()
                .unwrap_or(0)
        })
        .collect::<Vec<_>>();

    let mut lines = vec![
        "# zstd-rs Benchmark".to_string(),
        String::new(),
        format!("Source CSV: `{}`", csv_path.display()),
        String::new(),
    ];
    if !commentary.trim().is_empty() {
        lines.push(format!("Commentary: {}", commentary.trim()));
        lines.push(String::new());
    }
    lines.push("Percent improvements compare new/current against upstream. Each compressed output is decoded with C zstd and byte-compared against the original fixture.".to_string());
    lines.push(String::new());
    lines.push("```text".to_string());
    lines.push(format_row(&headers, &widths));
    let separators = widths
        .iter()
        .map(|width| "-".repeat(*width))
        .collect::<Vec<_>>();
    let separator_refs = separators.iter().map(String::as_str).collect::<Vec<_>>();
    lines.push(format_row(&separator_refs, &widths));
    for row in &table_rows {
        let row_refs = row.iter().map(String::as_str).collect::<Vec<_>>();
        lines.push(format_row(&row_refs, &widths));
    }
    lines.push("```".to_string());
    lines.push(String::new());
    write_all(path, &lines.join("\n"))
}

fn find_row<'a>(rows: &'a [Row], fixture: &str, encoder: &str) -> io::Result<&'a Row> {
    rows.iter()
        .find(|row| row.fixture == fixture && row.encoder == encoder)
        .ok_or_else(|| io::Error::other(format!("missing {encoder} result for {fixture}")))
}

fn format_row(row: &[&str], widths: &[usize]) -> String {
    row.iter()
        .enumerate()
        .map(|(idx, value)| format!("{value:<width$}", width = widths[idx]))
        .collect::<Vec<_>>()
        .join("  ")
}

fn format_number(value: u64) -> String {
    let text = value.to_string();
    let mut out = String::new();
    for (idx, ch) in text.chars().rev().enumerate() {
        if idx > 0 && idx % 3 == 0 {
            out.push(',');
        }
        out.push(ch);
    }
    out.chars().rev().collect()
}
